components/ucd/category/src/lib.rs

```diff
@@ -0,0 +1,313 @@
+        [UppercaseLetter, LowercaseLetter, TitlecaseLetter, ModifierLetter, OtherLetter]
+            .contains(self)
+    }
+
+    /// Mn | Mc | Me
+    ///
+    /// Abbreviated: M
+    pub fn is_mark(&self) -> bool {
+        [NonspacingMark, SpacingMark, EnclosingMark].contains(self)
+    }
+
+    /// Nd | Nl | No
+    ///
+    /// Abbreviated: N
+    pub fn is_number(&self) -> bool {
+        [DecimalNumber, LetterNumber, OtherNumber].contains(self)
+    }
+
+    /// Pc | Pd | Ps | Pe | Pi | Pf | Po
+    ///
+    /// Abbreviated: P
+    pub fn is_punctuation(&self) -> bool {
+        [
+            ConnectorPunctuation,
+            DashPunctuation,
+            OpenPunctuation,
+            ClosePunctuation,
+            InitialPunctuation,
+            FinalPunctuation,
+            OtherPunctuation,
+        ].contains(self)
+    }
+
+    /// Sm | Sc | Sk | So
+    ///
+    /// Abbreviated: S
+    pub fn is_symbol(&self) -> bool {
+        [MathSymbol, CurrencySymbol, ModifierLetter, OtherSymbol].contains(self)
+    }
+
+    /// Zs | Zl | Zp
+    ///
+    /// Abbreviated: Z
+    pub fn is_separator(&self) -> bool {
+        [SpaceSeparator, LineSeparator, ParagraphSeparator].contains(self)
+    }
+
+    /// Cc | Cf | Cs | Co | Cn
+    ///
+    /// Abbreviated: C
+    pub fn is_other(&self) -> bool {
+        [Control, Format, Surrogate, PrivateUse, Unassigned].contains(self)
+    }
+}
+
+const GENERAL_CATEGORY_TABLE: &'static [(char, char, GeneralCategory)] =
+    include!("tables/general_category.rsv");
+
+impl GeneralCategory {
+    /// Find the GeneralCategory of a single char.
+    pub fn of(ch: char) -> GeneralCategory {
+        bsearch_range_value_table(ch, GENERAL_CATEGORY_TABLE)
+    }
+}
+
+fn bsearch_range_value_table(
+    c: char,
+    r: &'static [(char, char, GeneralCategory)],
+) -> GeneralCategory {
+    match r.binary_search_by(|&(lo, hi, _)| if lo <= c && c <= hi {
+        Ordering::Equal
+    } else if hi < c {
+        Ordering::Less
+    } else {
+        Ordering::Greater
+    }) {
+        Ok(idx) => {
+            let (_, _, category) = r[idx];
+            category
+        }
+        Err(_) => GeneralCategory::Unassigned,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::GeneralCategory as GC;
+    use std::char;
+
+    #[test]
+    fn test_ascii() {
+        for c in 0..32 {
+            let c = char::from_u32(c).unwrap();
+            assert_eq!(GC::of(c), GC::Control);
+        }
+        assert_eq!(GC::of(' '), GC::SpaceSeparator);
+        assert_eq!(GC::of('!'), GC::OtherPunctuation);
+        assert_eq!(GC::of('"'), GC::OtherPunctuation);
+        assert_eq!(GC::of('#'), GC::OtherPunctuation);
+        assert_eq!(GC::of('$'), GC::CurrencySymbol);
+        assert_eq!(GC::of('%'), GC::OtherPunctuation);
+        assert_eq!(GC::of('&'), GC::OtherPunctuation);
+        assert_eq!(GC::of('\''), GC::OtherPunctuation);
+        // TODO I would like to do the rest of ASCII
```

Comment:
Yeah, corner cases of ASCII would be great, like U+0000, U+007F, and also edges of BMP (U+FFFF, U+FFFE, U+FFEF), and edges of PUA planes, specially (U+10FFFF).